### ISSUE TITLE
pm2 hangs when trying to run pm2 with non-default PM2_HOME and with another pm2 daemon running

### ISSUE THREAD

**Issue by malinoff** [Unitech/pm2#1434]
This can be easily reproduced with the following `test.js`:

``` javascript
var express = require('express');

var server = express();

server.all('/', function(req, res) {
    res.sendStatus(200);
})
var http = require('http');
http.createServer(server).listen(1234, function(){console.log('Started!');})
```

``` bash
# Run first pm2 with default PM2_HOME
$ ./node_modules/.bin/pm2 start test.js
[PM2] Spawning PM2 daemon
[PM2] PM2 Successfully daemonized
[PM2] Starting test.js in fork_mode (1 instance)
[PM2] Done.
┌──────────┬────┬──────┬───────┬────────┬─────────┬────────┬─────────────┬──────────┐
│ App name │ id │ mode │ pid   │ status │ restart │ uptime │ memory      │ watching │
├──────────┼────┼──────┼───────┼────────┼─────────┼────────┼─────────────┼──────────┤
│ test     │ 0  │ fork │ 30642 │ online │ 0       │ 0s     │ 17.477 MB   │ disabled │
└──────────┴────┴──────┴───────┴────────┴─────────┴────────┴─────────────┴──────────┘
 Use `pm2 show <id|name>` to get more details about an app

# Run second pm2 with non-default PM2_HOME
$ export PM2_HOME=./.pm2; mkdir ./.pm2; ./node_modules/.bin/pm2 start test.js
[PM2] Spawning PM2 daemon
# ... and nothing happens.
```

P.S. It is not caused by port collision, if you thought about it. The same behavior can be observed with different ports.

**Comment by soyuka** [comment]
Seems logic to me. Just kill pm2 before launching it with a new home.

**Comment by malinoff** [comment]
@soyuka I want to have multiple pm2 daemons running, 1 per project. I thought it inspects its current home and checks pidfile, no?

**Comment by soyuka** [comment]
As far as I know pm2 can't work like this. I've manage to do this by using one pm2 per unix-user. If you find a solution it can be interesting, let me know.

**Comment by malinoff** [comment]
@soyuka I don't know js so it would be difficult to me to investigate into the code. It would be awesome if you give me a starting point :)

**Comment by soyuka** [comment]
I think you'd have to change the RPC port constant too:
- [the rpc client](https://github.com/Unitech/PM2/blob/1b086fa6b31bbe8d311838ba77c401d98074ec6d/lib/Satan.js#L445) is started with the `DAEMON_RPC_PORT`. 
- [the interactor daemon](https://github.com/Unitech/PM2/blob/1b086fa6b31bbe8d311838ba77c401d98074ec6d/lib/Interactor/InteractorDaemonizer.js#L139) also has an `INTERACTOR_RPC_PORT` that would have to be unique. 

Those variables can be changed in the [`constant.js`](https://github.com/Unitech/PM2/blob/1b086fa6b31bbe8d311838ba77c401d98074ec6d/constants.js) file.

**Comment by malinoff** [comment]
I don't think that modifying something installed in `node_modules` is good idea. Also it looks like these parameters are internal and I think that pm2 should take care about that by picking the next available port if default is taken.

**Comment by malinoff** [comment]
P.S. Take a look at vagrant, how it manages ssh ports. As vagrant user I shouldn't care about them at all. If standard `2222` port is taken, `2200` will be used. If `2200` is taken, something else will be used. The point is that the whole process is fully automatic

**Comment by soyuka** [comment]
It's located in `~/.pm2/conf.js` by default, so it will be added to your pm2 home if not provided.

I don't think we want to overload pm2 with those kind of features.

**Comment by malinoff** [comment]
Ah, so I need to specify `PM2_HOME` AND put the variables into custom `$PM2_HOME/.pm2/conf.js`?

Such feature won't 'overload' pm2. It can be easily implemented with while loop, catching `EADDRINUSE` and incrementing integers.

**Comment by soyuka** [comment]
Omg no, this is not a clean solution ^^. I've developped [this module](https://github.com/soyuka/local-port) that could do the job. But I don't see why we would need this, your case is a bit particular, there is usually only one pm2 instance per server.

**Comment by malinoff** [comment]
Your module essentially does exactly what I described :) but let's stop the discussion about this.

I basically don't like global objects, and I like built-in mechanisms to do things.

I always write `init.d/` scripts/`systemd` service files for my services, and each service (obviously) can be started/restarted/stopped separately from any other service. When using PM2 this is quite difficult to achieve. So I have the following concerns:
1)`pm2 startup` command generates startup script for pm2 itself, not for services running within
2) this script relies on dumping & restoring services, implying that services should be registered before the startup script can be actually used
3) this script is global which means it cannot be used to stop one particular service
4) even if I write a custom startup script with `pm2 start script.js --name myservice` and `pm2 stop myservice`, since pm2 is global anybody can do `pm2 kill` (e.g. by using global startup script) which will kill my service too

In other words, running multiple pm2 instances is basically separation of responsibility. Each project is deployed into its own directory and has its own startup script. PM2 makes this setup literally impossible.

**Comment by soyuka** [comment]
> since pm2 is global anybody can do `pm2 kill`

Only the user that ran `pm2` would be able to do that

PM2 should be used as "global" and then run every instances from 1 instance. pm2 is a process manager and is used to manage services. By tweaking a little bit you'd be able to do the way you want - using 1 pm2 by service.

Have you manage to build a proof of concept by changing the constants I told you about?
Do you think that using some random port (for example with a module like mine) would help you to do that?

**Comment by Unitech** [comment]
Let's say I have a child.js file, now I want two instances of PM2, each one runing the child.js script:

```
# this will create the .pm2 in the home of current user then start echo.js
$ ./node_modules/.bin/pm2 start echo.js  
$ ./node_modules/.bin/pm2 ls

# this will create the .pm3 in the current folder then start echo.js
$ PM2_HOME='.pm3' ./node_modules/.bin/pm2 start echo.js 
$ PM2_HOME='.pm3' ./node_modules/.bin/pm2 ls
```

If your command is not responsive please do:

```
$ PM2_HOME='./.pm2' DEBUG="*" ./node_modules/.bin/pm2 start test.js
```

And paste here the output displayed by the command with DEBUG activated
